## Guided Project - Hacker News 

#### Hacker News posts from the past 12 months (including # of votes and comments)

This [data set](https://www.kaggle.com/hacker-news/hacker-news-posts#HN_posts_year_to_Sep_26_2016.csv) is Hacker News posts from the last 12 months (up to September 26 2016).

It includes the following columns:

* `title`: title of the post (self explanatory)

* `url`: the url of the item being linked to

* `num_points`: the number of upvotes the post received

* `num_comments`: the number of comments the post received

* `author`: the name of the account that made the post

* `created_at`: the date and time the post was made (the time zone is Eastern Time in the US)


In this project we are specifically interested in posts whose titles begin with either `Ask HN` or `Show HN`. Users submit Ask HN posts to ask the Hacker News community a specific question. Likewise, users submit `Show HN` posts to show the Hacker News community a project, product, or just generally something interesting.

We'll compare these two types of posts to determine the following:

* Do `Ask HN` or `Show HN` receive more comments on average?
* Do posts created at a certain time receive more comments on average?


### Read the file 
Read the file in as a list of lists and extract headers

In [1]:
from csv import reader
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)
headers = hn[0]
hn= hn[1:]
print(headers)
print(hn[:5])


['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


### Separating the posts - Ask HN and Show HN posts

In [2]:
ask_posts =[]
show_posts =[]
other_posts =[]
for i in hn:
    title = i[1]
    if title.lower().startswith('ask hn'):
        ask_posts.append(i)
    elif title.lower().startswith('show hn'):
        show_posts.append(i)
    else:
        other_posts.append(i)
ask_count = len(ask_posts)
show_count = len(show_posts)
others_count = len(other_posts)
print(" No. of ask posts:",ask_count,'\n',"No. of show posts:", show_count,'\n',"No. of other posts:", others_count)

 No. of ask posts: 1744 
 No. of show posts: 1162 
 No. of other posts: 17194


### Determine the average of comments in Ask HN and Show HN

In [3]:
## ask-posts
total_ask_comments = 0
for posts in ask_posts:
    num_comments = int(posts[4]) # convert to int to ease the calculation
    total_ask_comments += num_comments
avg_ask_comments = total_ask_comments/ask_count
print("Average comments in ask posts:",avg_ask_comments)

## show-posts
total_show_comments = 0
for posts in show_posts:
    num_comments = int(posts[4]) # convert to int to ease the calculation
    total_show_comments += num_comments
avg_show_comments = total_show_comments/show_count
print("Average comments in show posts:",avg_show_comments)


    

Average comments in ask posts: 14.038417431192661
Average comments in show posts: 10.31669535283993



From the average values as seen above, we can conclude that on an average the ask posts received approx. 14 comments more  than the show posts in the HN community.

### Finding  the number of posts and comments created in a certain time (By Hour)

Next, we'll determine if ask posts created at a certain time are more likely to attract comments. We'll use the following steps to perform this analysis:

1. Calculate the amount of ask posts created in each hour of the day, along with the number of comments received.

1. Calculate the average number of comments ask posts receive by hour created.


In [4]:
# Calculate the amount of ask posts created during 
# each hour of day and the number of comments received.

import datetime as dt
result_list = []
counts_by_hour={}
comments_by_hour ={}
date_format = "%m/%d/%Y %H:%M"
for posts in ask_posts:
    result_list.append(
        [posts[6], int(posts[4])]
    )

for posts in result_list:
    created_at = posts[0]
    comments = int(posts[1])
    hour = dt.datetime.strptime(created_at, date_format).strftime("%H")
    if hour in counts_by_hour:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += comments
    else:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = comments
print("Comments by hour \n",comments_by_hour)

print("\nCount by hour \n",counts_by_hour)
        
    

Comments by hour 
 {'17': 1146, '23': 543, '18': 1439, '20': 1722, '21': 1745, '14': 1416, '04': 337, '07': 267, '01': 683, '02': 1381, '15': 4477, '05': 464, '16': 1814, '19': 1188, '22': 479, '06': 397, '00': 447, '12': 687, '11': 641, '09': 251, '08': 492, '03': 421, '13': 1253, '10': 793}

Count by hour 
 {'17': 100, '23': 68, '18': 109, '20': 80, '21': 109, '14': 107, '04': 47, '07': 34, '01': 60, '02': 58, '15': 116, '05': 46, '16': 108, '19': 110, '22': 71, '06': 44, '00': 55, '12': 73, '11': 58, '09': 45, '08': 48, '03': 54, '13': 85, '10': 59}


### Calculate the average number of comments created for `Ask HN` posts during each hour of the day



In [5]:
# Calculate the average amount of comments 
# ask posts created at each hour of the day.
avg_by_hour=[]
for hour in comments_by_hour:
    avg_by_hour.append([hour,comments_by_hour[hour]/counts_by_hour[hour]])
# print(avg_by_hour)
avg_by_hour

[['17', 11.46],
 ['23', 7.985294117647059],
 ['18', 13.20183486238532],
 ['20', 21.525],
 ['21', 16.009174311926607],
 ['14', 13.233644859813085],
 ['04', 7.170212765957447],
 ['07', 7.852941176470588],
 ['01', 11.383333333333333],
 ['02', 23.810344827586206],
 ['15', 38.5948275862069],
 ['05', 10.08695652173913],
 ['16', 16.796296296296298],
 ['19', 10.8],
 ['22', 6.746478873239437],
 ['06', 9.022727272727273],
 ['00', 8.127272727272727],
 ['12', 9.41095890410959],
 ['11', 11.051724137931034],
 ['09', 5.5777777777777775],
 ['08', 10.25],
 ['03', 7.796296296296297],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696]]

### Sorting and displaying top 5 hours for ask post comments

In [14]:
swap_avg_by_hour = []

for i in avg_by_hour:
    swap_avg_by_hour.append([i[1],i[0]])
#print(swap_avg_by_hour)
 
sorted_swap = sorted(swap_avg_by_hour, reverse = True)

print("Top 5 Hours for `Ask` Posts Comments \n")
for avg, hr in sorted_swap[:5]:
    
    print("{}:00 {:.2f} average comments per post".format(hr,avg))
    



Top 5 Hours for `Ask` Posts Comments 

15:00 38.59 average comments per post
02:00 23.81 average comments per post
20:00 21.52 average comments per post
16:00 16.80 average comments per post
21:00 16.01 average comments per post


We can infer that, the posts created at 15:00hrs received an average of 38.59 comments per post.According to the [documentation](https://www.kaggle.com/hacker-news/hacker-news-posts), the time zone is Eastern Time in the US ie, posts created at 3:00PM EST have a higher chance of receiving more comments.

### Conclusion


In this project, wew worked with the [data set](https://www.kaggle.com/hacker-news/hacker-news-posts) of submissions to 
[Hacker News](https://news.ycombinator.com/) where user posts are voted and commented upon. We were comparing two types of posts: `Ask HN`- where users posted questions to Hacker News community  and `Show HN` - displays from the community. 

It was found out that on an average the ask posts received approx. 14 comments more than the show posts and posts created at 15:00hrs (3:00PM EST) received more comments. From our analysis, we can conclude that to receive more comments, posts should be created at 3:00PM EST.